com.unity.ml-agents.extensions/Runtime/Sensors/GridSensorBase.cs

```diff
@@ -0,0 +1,336 @@
+        ///         {
+        ///             channelValues[1] = goRb.velocity.x;
+        ///             channelValues[2] = goRb.velocity.y;
+        ///             channelValues[3] = goRb.velocity.z;
+        ///         }
+        ///         return channelValues;
+        ///     }
+        ///  </code>
+        /// </example>
+        protected virtual void GetObjectData(GameObject currentColliderGo, int typeIndex, float[] dataBuffer)
+        {
+            dataBuffer[0] = typeIndex;
+        }
+
+        /// <summary>
+        /// Get the observation size for each cell. This will be the size of dataBuffer in GetObjectData().
+        /// </summary>
+        protected virtual int GetObservationSize()
+        {
+            return 1;
+        }
+
+        /// <summary>
+        /// Whether the data is normailzed within [0, 1]. The sensor can only use PNG compression if the data is normailzed.
+        /// </summary>
+        protected virtual bool IsDataNormalized()
+        {
+            return false;
+        }
+
+        /// <summary>
+        /// Runs basic validation assertions to check that the values can be normalized
+        /// </summary>
+        /// <param name="channelValues">The values to be validated</param>
+        /// <param name="currentColliderGo">The gameobject used for better error messages</param>
+        protected virtual void ValidateValues(float[] channelValues, GameObject currentColliderGo)
+        {
+            if (m_CompressionType != SensorCompressionType.PNG)
+            {
+                return;
+            }
+
+            for (int j = 0; j < channelValues.Length; j++)
+            {
+                if (channelValues[j] < 0 || channelValues[j] > 1)
+                    throw new UnityAgentsException($"When using compression type {m_CompressionType} the data value has to be normalized between 0-1. " +
+                        $"Received value[{channelValues[j]}] for {currentColliderGo.name}");
+            }
+        }
+
+        /// <summary>
+        /// LoadObjectData - If the GameObject matches a tag, GetObjectData is called to extract the data from the GameObject
+        /// then the data is transformed based on the GridDepthType of the gridsensor.
+        /// Further documetation on the GridDepthType can be found below
+        /// </summary>
+        /// <param name="currentColliderGo">The game object that was found colliding with a certain cell</param>
+        /// <param name="cellIndex">The index of the current cell</param>
+        protected internal void LoadObjectData(GameObject currentColliderGo, int cellIndex)
+        {
+            Profiler.BeginSample("GridSensor.LoadObjectData");
+            for (var i = 0; i < m_DetectableObjects.Length; i++)
+            {
+                if (!ReferenceEquals(currentColliderGo, null) && currentColliderGo.CompareTag(m_DetectableObjects[i]))
+                {
+                    Array.Clear(m_CellDataBuffer, 0, m_CellDataBuffer.Length);
```

Comment:
This feels like a a good usecase for an ArraySegment that's a "view' on the part of m_CellDataBuffer that we're able to write to. Now that we're on 2019.4, we might be able to use a new enough version of C# that we can write, otherwise we could roll our own small version